src/finance.module/finance.service/finance.service.ts

```
@@ -0,0 +1,92 @@
+	}
+
+	public setActiveProviderByName(name: string) {
+		const key = Object.keys(providersNamesDict).find(
+			(key: string) => providersNamesDict[key as keyof typeof providersNamesDict] === name,
+		);
+		return this.setActiveProviderByKey(
+			providersKeysDict[key as keyof typeof providersKeysDict],
+		);
+	}
+
+	public setActiveProviderByKey(key: string) {
+		if (!Object.values(providersKeysDict).includes(key)) {
+			return false;
+		}
+		this._activeProviderKey = key;
+		this._cacheManager.reset();
+		return true;
+	}
+
+	public async getBtcUahRateAsync(): Promise<number> {
+		const cacheName = 'BTC_UAH_RATE';
+		if (this.cacheRateConfig.cacheActive && (await this._cacheManager.get(cacheName))) {
+			this.logger.debug(`Cache used for request ${cacheName}`);
+			return Number(await this._cacheManager.get(cacheName));
+		}
+		let providerInstance = this._financeProviderFabric.getProviderByKey(
+			this._activeProviderKey,
+		);
+		let rateValue = 0;
+		try {
+			rateValue = await this.getBtcUahRateFromProviderAsync(providerInstance);
+		} catch (e) {
+			this.logger.error(
+				`Unable to get rate value from active provider. AutoSwitch: ${
+					this.autoChangeUnavailableProviders ? 'Enabled' : 'Disabled'
+				}`,
+			);
+			if (this.autoChangeUnavailableProviders) {
+				for (const key in providersKeysDict) {
+					const providerKey = providersKeysDict[key as keyof typeof providersKeysDict];
+					if (providerKey !== this._activeProviderKey) {
+						providerInstance = this._financeProviderFabric.getProviderByKey(
+							providersKeysDict[key as keyof typeof providersKeysDict],
+						);
+						try {
+							rateValue = await this.getBtcUahRateFromProviderAsync(providerInstance);
+							break;
+						} catch (e) {
+							this.logger.error(
+								`Unable to get rate value from ${providerInstance.providerName} provider`,
+							);
+						}
+					}
+				}
+			}
+		}
+		if (!rateValue) throw new Error(rateErrorsDict.INVALID_RATE_VALUE);
```

Should probably throw the error from the last provider. That way the error would be more informative

---

You are right, but does the user need more information especially only from last provider? Everything writing in application logs with details

---

Okay, if you log this error this might not be needed.